File: apps/azimuth/src/routes/atlas.ts
import { type AtlasDb, atlasProjects, builds, commits, createAtlasDb, downloads, versions } from '@bx-team/stratus/d1';
import { BuildsQuerySchema, CreateVersionBodySchema, UploadMetadataSchema } from '@bx-team/types/schema/atlas';
import { and, desc, eq, inArray } from 'drizzle-orm';
import { Hono } from 'hono';
import { createMiddleware } from 'hono/factory';
import type { Env } from '../env';
import { groupVersions } from '../lib/versions';
import { edgeCache } from '../middleware/cache';

/**
 * Atlas/Downloads route group — project, version and build metadata backed by atlas-db
 * and the ATLAS_BUCKET R2 binding. Public GETs go through the edge cache.
 */

type AtlasEnv = { Bindings: Env };

type ProjectRow = typeof atlasProjects.$inferSelect;
type VersionRow = typeof versions.$inferSelect;
type BuildRow = typeof builds.$inferSelect;
type CommitRow = typeof commits.$inferSelect;
type DownloadRow = typeof downloads.$inferSelect;

const requireApiSecret = createMiddleware<AtlasEnv>(async (c, next) => {
  const auth = c.req.header('Authorization');
  if (!auth || auth !== `Bearer ${c.env.API_SECRET_KEY}`) {
    return c.json({ ok: false, error: 'Unauthorized', message: 'Unauthorized' }, 401);
  }
  await next();
});

function projectResponse(project: ProjectRow, versionKeys: string[]) {
  return {
    project: {
      id: project.key,
      name: project.name,
      ...(project.description && { description: project.description }),
      ...(project.latestVersion && { latestVersion: project.latestVersion }),
      ...(project.experimentalVersion && { experimentalVersion: project.experimentalVersion }),
    },
    version_groups: groupVersions(versionKeys),
  };
}

function versionResponse(version: VersionRow, buildNumbers: number[]) {
  return {
    version: {
      id: version.key,
      ...(version.javaMinVersion && { java: { version: { minimum: version.javaMinVersion } } }),
      support: { status: version.supportStatus },
    },
    builds: buildNumbers.sort((a, b) => b - a),
  };
}

function buildResponse(build: BuildRow, buildCommits: CommitRow[], buildDownloads: DownloadRow[], publicUrl: string) {
  const downloadsMap: Record<string, { name: string; checksums: { sha256: string }; size: number; url: string }> = {};
  for (const dl of buildDownloads) {
    downloadsMap[dl.name] = {
      name: dl.fileName,
      checksums: { sha256: dl.sha256 },
      size: dl.size,
      url: `${publicUrl}/${dl.filePath}`,
    };
  }

  return {
    id: build.buildNumber,
    time: build.time.toISOString(),
    channel: build.channel,
    commits: buildCommits.map(c => ({ sha: c.sha, message: c.message, time: c.time.toISOString() })),
    downloads: downloadsMap,
  };
}

async function findProject(db: AtlasDb, projectKey: string) {
  const [project] = await db.select().from(atlasProjects).where(eq(atlasProjects.key, projectKey)).limit(1);
  return project;
}

async function findVersion(db: AtlasDb, projectId: number, versionKey: string) {
  const [version] = await db
    .select()
    .from(versions)
    .where(and(eq(versions.projectId, projectId), eq(versions.key, versionKey)))
    .limit(1);
  return version;
}

/** Batched (non-N+1) commit/download lookups for a set of builds. */
async function buildDetails(db: AtlasDb, buildIds: number[]) {
  const [allCommits, allDownloads] = buildIds.length
    ? await Promise.all([
        db.select().from(commits).where(inArray(commits.buildId, buildIds)),
        db.select().from(downloads).where(inArray(downloads.buildId, buildIds)),
      ])
    : [[], []];

  const commitsByBuild = new Map<number, CommitRow[]>();
  for (const commit of allCommits) {
    commitsByBuild.set(commit.buildId, [...(commitsByBuild.get(commit.buildId) ?? []), commit]);
  }
  const downloadsByBuild = new Map<number, DownloadRow[]>();
  for (const download of allDownloads) {
    downloadsByBuild.set(download.buildId, [...(downloadsByBuild.get(download.buildId) ?? []), download]);
  }

  return { commitsByBuild, downloadsByBuild };
}

export const atlas = new Hono<AtlasEnv>();

atlas.use('/*', edgeCache);

atlas.get('/projects', async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);

  try {
    const allProjects = await db.select().from(atlasProjects);
    const allVersions = allProjects.length
      ? await db
          .select()
          .from(versions)
          .where(
            inArray(
              versions.projectId,
              allProjects.map(p => p.id),
            ),
          )
      : [];

    const versionKeysByProject = new Map<number, string[]>();
    for (const version of allVersions) {
      versionKeysByProject.set(version.projectId, [
        ...(versionKeysByProject.get(version.projectId) ?? []),
        version.key,
      ]);
    }

    return c.json({
      projects: allProjects.map(project => projectResponse(project, versionKeysByProject.get(project.id) ?? [])),
    });
  } catch (error) {
    console.error('Error fetching projects:', error);
    return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to fetch projects' }, 500);
  }
});

atlas.get('/projects/:project', async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');

  try {
    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const projectVersions = await db.select().from(versions).where(eq(versions.projectId, project.id));

    return c.json(
      projectResponse(
        project,
        projectVersions.map(v => v.key),
      ),
    );
  } catch (error) {
    console.error('Error fetching project:', error);
    return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to fetch project' }, 500);
  }
});

atlas.get('/projects/:project/versions', async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');

  try {
    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const projectVersions = await db.select().from(versions).where(eq(versions.projectId, project.id));
    const versionBuilds = projectVersions.length
      ? await db
          .select()
          .from(builds)
          .where(
            inArray(
              builds.versionId,
              projectVersions.map(v => v.id),
            ),
          )
      : [];

    const buildNumbersByVersion = new Map<number, number[]>();
    for (const build of versionBuilds) {
      buildNumbersByVersion.set(build.versionId, [
        ...(buildNumbersByVersion.get(build.versionId) ?? []),
        build.buildNumber,
      ]);
    }

    return c.json(
      projectVersions.map(version => versionResponse(version, buildNumbersByVersion.get(version.id) ?? [])),
    );
  } catch (error) {
    console.error('Error fetching versions:', error);
    return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to fetch versions' }, 500);
  }
});

atlas.post('/projects/:project/versions/create', requireApiSecret, async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');

  try {
    const parsed = CreateVersionBodySchema.safeParse(await c.req.json().catch(() => ({})));
    if (!parsed.success) {
      return c.json({ ok: false, error: 'Bad Request', message: parsed.error.message }, 400);
    }
    const body = parsed.data;

    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const existingVersion = await findVersion(db, project.id, body.key);
    if (existingVersion) {
      return c.json(
        { ok: false, error: 'Conflict', message: `Version '${body.key}' already exists for project '${projectKey}'` },
        409,
      );
    }

    const [version] = await db
      .insert(versions)
      .values({
        projectId: project.id,
        key: body.key,
        supportStatus: body.supportStatus ?? 'SUPPORTED',
        javaMinVersion: body.javaMinVersion ?? null,
      })
      .returning();

    if (!version) {
      return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to create version' }, 500);
    }

    return c.json(
      {
        message: 'Version created successfully',
        version: {
          id: version.id,
          project: projectKey,
          key: version.key,
          supportStatus: version.supportStatus,
          javaMinVersion: version.javaMinVersion,
        },
      },
      201,
    );
  } catch (error) {
    console.error('Error creating version:', error);
    return c.json(
      {
        ok: false,
        error: 'Internal Server Error',
        message: `Failed to create version: ${error instanceof Error ? error.message : 'Unknown error'}`,
      },
      500,
    );
  }
});

atlas.get('/projects/:project/versions/:version', async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');
  const versionKey = c.req.param('version');

  try {
    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const version = await findVersion(db, project.id, versionKey);
    if (!version) {
      return c.json(
        { ok: false, error: 'Not Found', message: `Version '${versionKey}' not found for project '${projectKey}'` },
        404,
      );
    }

    const versionBuilds = await db.select().from(builds).where(eq(builds.versionId, version.id));

    return c.json(
      versionResponse(
        version,
        versionBuilds.map(b => b.buildNumber),
      ),
    );
  } catch (error) {
    console.error('Error fetching version:', error);
    return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to fetch version' }, 500);
  }
});

atlas.get('/projects/:project/versions/:version/builds', async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');
  const versionKey = c.req.param('version');

  const parsedQuery = BuildsQuerySchema.safeParse(c.req.query());
  if (!parsedQuery.success) {
    return c.json({ ok: false, error: 'Bad Request', message: parsedQuery.error.message }, 400);
  }
  const { channel: channelFilter } = parsedQuery.data;

  try {
    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const version = await findVersion(db, project.id, versionKey);
    if (!version) {
      return c.json(
        { ok: false, error: 'Not Found', message: `Version '${versionKey}' not found for project '${projectKey}'` },
        404,
      );
    }

    const conditions = [eq(builds.versionId, version.id)];
    if (channelFilter) conditions.push(eq(builds.channel, channelFilter));

    const versionBuilds = await db
      .select()
      .from(builds)
      .where(and(...conditions))
      .orderBy(desc(builds.buildNumber));

    const { commitsByBuild, downloadsByBuild } = await buildDetails(
      db,
      versionBuilds.map(b => b.id),
    );

    return c.json(
      versionBuilds.map(build =>
        buildResponse(
          build,
          commitsByBuild.get(build.id) ?? [],
          downloadsByBuild.get(build.id) ?? [],
          c.env.R2_PUBLIC_URL,
        ),
      ),
    );
  } catch (error) {
    console.error('Error fetching builds:', error);
    return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to fetch builds' }, 500);
  }
});

atlas.get('/projects/:project/versions/:version/builds/latest', async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');
  const versionKey = c.req.param('version');

  try {
    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const version = await findVersion(db, project.id, versionKey);
    if (!version) {
      return c.json(
        { ok: false, error: 'Not Found', message: `Version '${versionKey}' not found for project '${projectKey}'` },
        404,
      );
    }

    const [build] = await db
      .select()
      .from(builds)
      .where(eq(builds.versionId, version.id))
      .orderBy(desc(builds.buildNumber))
      .limit(1);

    if (!build) {
      return c.json(
        {
          ok: false,
          error: 'Not Found',
          message: `No builds found for version '${versionKey}' of project '${projectKey}'`,
        },
        404,
      );
    }

    const { commitsByBuild, downloadsByBuild } = await buildDetails(db, [build.id]);

    return c.json(
      buildResponse(
        build,
        commitsByBuild.get(build.id) ?? [],
        downloadsByBuild.get(build.id) ?? [],
        c.env.R2_PUBLIC_URL,
      ),
    );
  } catch (error) {
    console.error('Error fetching latest build:', error);
    return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to fetch latest build' }, 500);
  }
});

atlas.get('/projects/:project/versions/:version/builds/:build', async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');
  const versionKey = c.req.param('version');
  const buildParam = c.req.param('build');

  if (!/^\d+$/.test(buildParam)) {
    return c.json({ ok: false, error: 'Bad Request', message: 'Build number must be a positive integer' }, 400);
  }
  const buildNumber = parseInt(buildParam, 10);

  try {
    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const version = await findVersion(db, project.id, versionKey);
    if (!version) {
      return c.json(
        { ok: false, error: 'Not Found', message: `Version '${versionKey}' not found for project '${projectKey}'` },
        404,
      );
    }

    const [build] = await db
      .select()
      .from(builds)
      .where(and(eq(builds.versionId, version.id), eq(builds.buildNumber, buildNumber)))
      .limit(1);

    if (!build) {
      return c.json(
        {
          ok: false,
          error: 'Not Found',
          message: `Build '${buildNumber}' not found for version '${versionKey}' of project '${projectKey}'`,
        },
        404,
      );
    }

    const { commitsByBuild, downloadsByBuild } = await buildDetails(db, [build.id]);

    return c.json(
      buildResponse(
        build,
        commitsByBuild.get(build.id) ?? [],
        downloadsByBuild.get(build.id) ?? [],
        c.env.R2_PUBLIC_URL,
      ),
    );
  } catch (error) {
    console.error('Error fetching build:', error);
    return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to fetch build' }, 500);
  }
});

atlas.post('/projects/:project/versions/:version/builds/upload', requireApiSecret, async c => {
  const db = createAtlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');
  const versionKey = c.req.param('version');

  try {
    const body = await c.req.parseBody();
    const filePart = body.file;

    if (!(filePart instanceof File)) {
      return c.json({ ok: false, error: 'Bad Request', message: 'No file provided' }, 400);
    }

    const textPart = (name: string) => {
      const part = body[name];
      return typeof part === 'string' && part.length ? part : undefined;
    };

    const metadataRaw = textPart('metadata');
    const metadataInput: Record<string, unknown> = metadataRaw ? JSON.parse(metadataRaw) : {};

    if (metadataInput.buildNumber === undefined) {
      const raw = textPart('buildNumber');
      if (raw !== undefined) metadataInput.buildNumber = Number(raw);
    }
    if (metadataInput.channel === undefined) {
      const raw = textPart('channel');
      if (raw !== undefined) metadataInput.channel = raw;
    }
    if (metadataInput.commits === undefined) {
      const raw = textPart('commits');
      if (raw !== undefined) metadataInput.commits = JSON.parse(raw);
    }

    const metadata = UploadMetadataSchema.parse(metadataInput);

    const project = await findProject(db, projectKey);
    if (!project) {
      return c.json({ ok: false, error: 'Not Found', message: `Project '${projectKey}' not found` }, 404);
    }

    const version = await findVersion(db, project.id, versionKey);
    if (!version) {
      return c.json(
        { ok: false, error: 'Not Found', message: `Version '${versionKey}' not found for project '${projectKey}'` },
        404,
      );
    }

    let buildNumber = metadata.buildNumber;
    if (!buildNumber) {
      const [lastBuild] = await db
        .select()
        .from(builds)
        .where(eq(builds.versionId, version.id))
        .orderBy(desc(builds.buildNumber))
        .limit(1);
      buildNumber = lastBuild ? lastBuild.buildNumber + 1 : 1;
    }

    const fileBytes = new Uint8Array(await filePart.arrayBuffer());
    const fileName = filePart.name || 'upload.jar';
    const r2Key = `${projectKey}/versions/${versionKey}/${buildNumber}/${fileName}`;

    const hashBuffer = await crypto.subtle.digest('SHA-256', fileBytes);
    const sha256 = Array.from(new Uint8Array(hashBuffer))
      .map(b => b.toString(16).padStart(2, '0'))
      .join('');

    const channelValue = (metadata.channel?.toUpperCase() || 'STABLE') as 'ALPHA' | 'BETA' | 'STABLE';

    await c.env.ATLAS_BUCKET.put(r2Key, fileBytes, {
      httpMetadata: {
        contentType: fileName.endsWith('.jar') ? 'application/java-archive' : 'application/octet-stream',
      },
    });

    const [created] = await db
      .insert(builds)
      .values({ versionId: version.id, buildNumber, channel: channelValue, time: new Date() })
      .returning();

    if (!created) {
      return c.json({ ok: false, error: 'Internal Server Error', message: 'Failed to insert build' }, 500);
    }

    // D1 has no interactive transactions; db.batch() is the atomic unit, and the
    // build row is compensated away if the child inserts fail.
    try {
      const downloadInsert = db.insert(downloads).values({
        buildId: created.id,
        name: 'application',
        fileName,
        filePath: r2Key,
        size: fileBytes.length,
        sha256,
      });

      if (metadata.commits?.length) {
        await db.batch([
          db.insert(commits).values(
            metadata.commits.map(commit => ({
              buildId: created.id,
              sha: commit.sha,
              message: commit.message,
              time: new Date(commit.time),
            })),
          ),
          downloadInsert,
        ]);
      } else {
        await downloadInsert;
      }
    } catch (error) {
      await db.delete(builds).where(eq(builds.id, created.id));
      throw error;
    }

    return c.json({
      message: 'Build uploaded successfully',
      build: {
        id: buildNumber,
        project: projectKey,
        version: versionKey,
        channel: channelValue,
      },
    });
  } catch (error) {
    console.error('Error uploading build:', error);
    return c.json(
      {
        ok: false,
        error: 'Internal Server Error',
        message: `Failed to upload build: ${error instanceof Error ? error.message : 'Unknown error'}`,
      },
      500,
    );
  }
});

File: packages/database/src/models/atlas.rs
use chrono::{DateTime, Utc};

use crate::{Db, Error, Transaction};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Project {
    pub id: i64,
    pub key: String,
    pub name: String,
    pub description: Option<String>,
    pub latest_version: Option<String>,
    pub experimental_version: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Version {
    pub id: i64,
    pub project_id: i64,
    pub key: String,
    pub support_status: String,
    pub java_min_version: Option<i32>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Build {
    pub id: i64,
    pub version_id: i64,
    pub build_number: i64,
    pub channel: String,
    pub time: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Commit {
    pub build_id: i64,
    pub sha: String,
    pub message: String,
    pub time: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Download {
    pub build_id: i64,
    pub name: String,
    pub file_name: String,
    pub file_path: String,
    pub size: i64,
    pub sha256: String,
}

pub async fn projects(db: &Db) -> Result<Vec<Project>, Error> {
    sqlx::query_as!(
        Project,
        "SELECT id, key, name, description, latest_version, experimental_version
           FROM atlas.projects ORDER BY key",
    )
    .fetch_all(db)
    .await
}

pub async fn project(db: &Db, key: &str) -> Result<Option<Project>, Error> {
    sqlx::query_as!(
        Project,
        "SELECT id, key, name, description, latest_version, experimental_version
           FROM atlas.projects WHERE key = $1",
        key,
    )
    .fetch_optional(db)
    .await
}

pub async fn create_project(
    db: &Db,
    key: &str,
    name: &str,
    description: Option<&str>,
) -> Result<Project, Error> {
    sqlx::query_as!(
        Project,
        "INSERT INTO atlas.projects (key, name, description) VALUES ($1, $2, $3)
         RETURNING id, key, name, description, latest_version, experimental_version",
        key,
        name,
        description,
    )
    .fetch_one(db)
    .await
}

pub async fn versions_of(db: &Db, project_ids: &[i64]) -> Result<Vec<Version>, Error> {
    sqlx::query_as!(
        Version,
        "SELECT id, project_id, key, support_status, java_min_version
           FROM atlas.versions WHERE project_id = ANY($1)",
        project_ids,
    )
    .fetch_all(db)
    .await
}

pub async fn version(db: &Db, project_id: i64, key: &str) -> Result<Option<Version>, Error> {
    sqlx::query_as!(
        Version,
        "SELECT id, project_id, key, support_status, java_min_version
           FROM atlas.versions WHERE project_id = $1 AND key = $2",
        project_id,
        key,
    )
    .fetch_optional(db)
    .await
}

pub async fn create_version(
    db: &Db,
    project_id: i64,
    key: &str,
    support_status: &str,
    java_min_version: Option<i32>,
) -> Result<Version, Error> {
    sqlx::query_as!(
        Version,
        "INSERT INTO atlas.versions (project_id, key, support_status, java_min_version)
         VALUES ($1, $2, $3, $4)
         RETURNING id, project_id, key, support_status, java_min_version",
        project_id,
        key,
        support_status,
        java_min_version,
    )
    .fetch_one(db)
    .await
}

pub async fn build_numbers(db: &Db, version_id: i64) -> Result<Vec<i64>, Error> {
    sqlx::query_scalar!(
        "SELECT build_number FROM atlas.builds WHERE version_id = $1 ORDER BY build_number DESC",
        version_id,
    )
    .fetch_all(db)
    .await
}

pub async fn builds_of(
    db: &Db,
    version_id: i64,
    channel: Option<&str>,
) -> Result<Vec<Build>, Error> {
    sqlx::query_as!(
        Build,
        "SELECT id, version_id, build_number, channel, time
           FROM atlas.builds
          WHERE version_id = $1 AND ($2::text IS NULL OR channel = $2)
          ORDER BY build_number DESC",
        version_id,
        channel,
    )
    .fetch_all(db)
    .await
}

pub async fn build(db: &Db, version_id: i64, build_number: i64) -> Result<Option<Build>, Error> {
    sqlx::query_as!(
        Build,
        "SELECT id, version_id, build_number, channel, time
           FROM atlas.builds WHERE version_id = $1 AND build_number = $2",
        version_id,
        build_number,
    )
    .fetch_optional(db)
    .await
}

pub async fn latest_build(db: &Db, version_id: i64) -> Result<Option<Build>, Error> {
    sqlx::query_as!(
        Build,
        "SELECT id, version_id, build_number, channel, time
           FROM atlas.builds WHERE version_id = $1
          ORDER BY build_number DESC LIMIT 1",
        version_id,
    )
    .fetch_optional(db)
    .await
}

pub async fn commits_of(db: &Db, build_ids: &[i64]) -> Result<Vec<Commit>, Error> {
    sqlx::query_as!(
        Commit,
        "SELECT build_id, sha, message, time FROM atlas.commits
          WHERE build_id = ANY($1) ORDER BY id",
        build_ids,
    )
    .fetch_all(db)
    .await
}

pub async fn downloads_of(db: &Db, build_ids: &[i64]) -> Result<Vec<Download>, Error> {
    sqlx::query_as!(
        Download,
        "SELECT build_id, name, file_name, file_path, size, sha256 FROM atlas.downloads
          WHERE build_id = ANY($1) ORDER BY id",
        build_ids,
    )
    .fetch_all(db)
    .await
}

/// Reserves the next build number for a version inside the caller's transaction.
///
/// The row lock, not a read-then-write, is what stops two CI jobs from claiming the same number.
pub async fn next_build_number(tx: &mut Transaction<'_>, version_id: i64) -> Result<i64, Error> {
    sqlx::query_scalar!(
        "SELECT id FROM atlas.versions WHERE id = $1 FOR UPDATE",
        version_id
    )
    .fetch_one(&mut **tx)
    .await?;

    sqlx::query_scalar!(
        "SELECT coalesce(max(build_number), 0) + 1 FROM atlas.builds WHERE version_id = $1",
        version_id,
    )
    .fetch_one(&mut **tx)
    .await
    .map(|value| value.unwrap_or(1))
}

pub async fn insert_build(
    tx: &mut Transaction<'_>,
    version_id: i64,
    build_number: i64,
    channel: &str,
    time: DateTime<Utc>,
) -> Result<Build, Error> {
    sqlx::query_as!(
        Build,
        "INSERT INTO atlas.builds (version_id, build_number, channel, time)
         VALUES ($1, $2, $3, $4)
         RETURNING id, version_id, build_number, channel, time",
        version_id,
        build_number,
        channel,
        time,
    )
    .fetch_one(&mut **tx)
    .await
}

pub async fn insert_commits(
    tx: &mut Transaction<'_>,
    build_id: i64,
    shas: &[String],
    messages: &[String],
    times: &[DateTime<Utc>],
) -> Result<(), Error> {
    if shas.is_empty() {
        return Ok(());
    }
    sqlx::query!(
        "INSERT INTO atlas.commits (build_id, sha, message, time)
         SELECT $1, sha, message, time
           FROM unnest($2::text[], $3::text[], $4::timestamptz[]) AS c(sha, message, time)",
        build_id,
        shas,
        messages,
        times,
    )
    .execute(&mut **tx)
    .await?;
    Ok(())
}

pub async fn insert_download(
    tx: &mut Transaction<'_>,
    build_id: i64,
    name: &str,
    file_name: &str,
    file_path: &str,
    size: i64,
    sha256: &str,
) -> Result<(), Error> {
    sqlx::query!(
        "INSERT INTO atlas.downloads (build_id, name, file_name, file_path, size, sha256)
         VALUES ($1, $2, $3, $4, $5, $6)",
        build_id,
        name,
        file_name,
        file_path,
        size,
        sha256,
    )
    .execute(&mut **tx)
    .await?;
    Ok(())
}
